message_actions: Add move message option

SilentCruzer opened the pull request:
Closes: part of #5165 

#### Changes:
- Created a new component called MoveMessage to display the details on where to move the message. The screen changes slightly depending on whether the user is admin or not. If the user is an admin, they can change the stream of the message, if not admin they can only change the topic in the current stream.
- Added navigation to the above screen.
- Updated the update message API, added a new param called stream_id and changes the type of propagate_mode from boolean to string according to the API documentation.
- Added the propagate_mode to allow users to choose the option to move one message/ later messages / previous and following messages.

#### Changes video:

https://user-images.githubusercontent.com/73246484/148679305-9b02f6f7-5e2a-4729-ba47-874c7facc33b.mp4

#### Admin user UI (the stream part becomes a drop down):
<img src="https://user-images.githubusercontent.com/73246484/148679389-2bab7231-141a-4d07-9ea7-59b246e32628.jpg" height=600 width=300> <img src="https://user-images.githubusercontent.com/73246484/148679396-25486fd4-219b-47df-9b4d-30a61c10ee33.jpg" height=600 width=300>

gnprice commented on src/message/moveMessage.js:
```diff
@@ -0,0 +1,225 @@
+    },
+  };
+
+  const updateTextInput = (textInput, text) => {
+    if (textInput === null) {
+      // Depending on the lifecycle events this function is called from,
+      // this might not be set yet.
+      return;
```
This comment appears to be copy-pasted from `ComposeBox.js`. Is the comment true here? What does it mean?

… Reading a bit more, it looks like in fact `textInput` will *always* be null here -- it comes from `topicInputRef.current`, and we never set `topicInputRef` or pass it somewhere that could end up setting it. So this function never does anything.

gnprice commented on src/action-sheets/index.js:
```diff
@@ -449,6 +456,9 @@ export const constructMessageActionButtons = ({
     && (isStreamOrTopicNarrow(narrow) || isPmNarrow(narrow))
   ) {
     buttons.push(editMessage);
+    moveMessage.narrow = narrow;
+    moveMessage.admin = ownUser.is_admin;
```
This appears to be setting properties on the function (which is a global constant for this module), in order to store some of the data that we're currently processing.

Do not do this. This is a brittle pattern -- effectively you're storing these values as global variables.

Instead, use the mechanisms that the surrounding code already uses to pass the data that comparable functions need. Here, `moveMessage` needs to be a `Button<MessageArgs>`, because that's what's in the array that `constructMessageActionButtons` returns. Looking at the definition of `Button<Args>`, in general the function takes an `Args` -- so for `Button<MessageArgs>` it takes a `MessageArgs`. So we just need the relevant information to be added to `MessageArgs`.

Specifically we need `narrow` there. For `ownUser.is_admin`, the `moveMessage` callback doesn't really need that information -- the `MoveMessage` component will need it, but it can get that information from Redux for itself.

gnprice replied:
BTW in general if you see a way to make the code work that involves structuring it with a bad pattern, and you can't find a cleaner way to do it, it's OK to send a PR where it uses the bad structure (especially when it's a small part of the PR's code, like here) -- *but* it's important to call attention to it clearly. Typically that might mean
* both a TODO comment in the code,
* and a prominent note about it in the PR description, or as a PR comment shortly after sending the PR.

That way you help the reviewer make sure to notice it so they can offer suggestions on how to handle the problem.

Better yet, typically, is to ask in chat about how to handle the thing you're trying to do, before you have the PR ready to send. But sometimes the easiest way to make the question clear is with the context of the rest of the code, and then a PR with clear communication about the question is a good way to go.

gnprice commented on src/message/moveMessage.js:
```diff
@@ -0,0 +1,225 @@
+  };
+
+  const updateMessage = async () => {
+    const stream_info = await api.getStreamId(auth, stream);
```
This will go and make a request to the server to look up what stream has the given name.

We don't want to do that. We already have data on all the streams the user has access to know about, and that's the data we used for presenting the UI the user will have used to select the stream. We should use that data instead.

That's better because not only does it make things faster, and more reliable on a weak network connection -- because we save a request round-trip -- but also it keeps the action we actually take aligned with the action the user chose, eliminating a class of bugs where there was a mismatch between the data we had at the moment the user chose something in the UI and the data the server has by the time it handles the request.

For example, if I go to move something to stream `#awesome-team`, and right around that moment someone else goes and renames `#awesome-team` to `#awesome team`, things should just work -- the messages should get moved to that stream, even though its name was `#awesome team` by the time the request made it to the server.

(This connects to the point above about using stream IDs rather than stream names.)

gnprice reviewed (commented):
Thanks @SilentCruzer! Here's a partial review, with the things I had time to get to tonight. (There's more logic I didn't get to, especially the bottom half of the `MoveMessage` component's definition.)

Please take a look and make changes, and don't hesitate to ask questions in `#mobile-team` for anything you're unsure about.

SilentCruzer commented on src/message/MoveMessage.js:
```diff
@@ -0,0 +1,216 @@
+    setNarrow(streamNarrow(String(pickedStreamName)));
+  };
+
+  const handlePropagateMode = (propagatePickerIndex: number) => {
```
So this function is to handle the propagate model while using the Picker component.

After I added the types for the propagate_mode in updateMessage API, the setPropagateMode function does not work with onValueChange function of the Picker component(line 202). 

Since there will be only three values, I used if-else statements based on the values of the index

gnprice replied:
This works, but is fragile: these conditions in `handlePropagateMode` now have to line up with the order of options in the picker, and they're separated by quite a bit of code. It'd be easy for them to get out of sync, so that the items no longer had the effects they're labelled for.

For that reason it's better to use the `value` feature of Picker / PickerItem.

If you get errors from Flow but you're confident that what actually happens with the data is correct, you can add a `$FlowFixMe` comment with an explanation of why the actual behavior is correct.

SilentCruzer commented on src/message/MoveMessage.js:
```diff
@@ -0,0 +1,216 @@
+    Toast.show('Moved Message');
+  };
+
+  const handleNarrow = (pickedStreamId: number) => {
```
Noticed a bug when I made the PR. When move message function was accessed through the topic narrow, the topic auto complete didn't suggest any topic. So I had to create a new stream narrow.

To create a stream narrow, I needed the name of the stream.

First idea was to use the label I added to the items in the Picker component, but I later realized that I can't access the label after adding it, so I used the below method.

Since I had the id of the stream, mapped through the current stream list, and found the name,

gnprice replied:
Cool, looking up the stream name given the stream ID is the right direction.

Don't do it with a linear search through the whole list of streams, though: #3339.

Instead, use `getStreamsByName` as I mentioned in another comment.

gnprice commented on src/message/MoveMessage.js:
```diff
@@ -0,0 +1,216 @@
+  const isAdmin = useSelector(getOwnUser).is_admin;
+  const messageId = props.route.params.message.id;
+  const currentStreamName = streamNameOfNarrow(props.route.params.messageNarrow);
+  const currentStreamId = allStreams.find(s => s.name === currentStreamName)?.stream_id;
```
Avoid linear searches like this one: #3339

That's why at https://github.com/zulip/zulip-mobile/pull/5189#discussion_r781797902 I recommended `getStreamsByName`, which is a data structure for accomplishing the same lookup efficiently.

gnprice commented on src/message/MoveMessage.js:
```diff
@@ -0,0 +1,216 @@
+
+  const updateMessage = () => {
+    if (isAdmin) {
+      api
+        .updateMessage(
+          auth,
+          { subject, stream_id: streamId, propagate_mode: propagateMode },
+          messageId,
+        )
+        .catch(error => {
```
Don't use `Promise#catch`; instead, make the surrounding function async, and use `await` inside a try/catch block. (You can search for `await api` to find a number of examples.)

The reason is that with `Promise#catch` it's hard to avoid having subtle edge cases where exceptions don't get handled properly, and using try/catch blocks instead gives a much more robust structure for managing our exception handling.

(This isn't something that's obvious a priori. I wish I had a good writeup of it to link you to, but I don't.)

gnprice replied:
Oh, and here's a related benefit of using try/catch: you have these two `.catch` calls that are completely identical, and it would be better to say that just once. With try/catch, that's very straightforward to do: just put the try/catch outside the if/else that controls the two different paths.

gnprice commented on src/message/MoveMessage.js:
```diff
@@ -0,0 +1,216 @@
+    } else {
+      api
+        .updateMessage(auth, { subject, propagate_mode: propagateMode }, messageId)
+        .catch(error => {
+          showErrorAlert('Failed to move message', error.message);
+          props.navigation.goBack();
+        });
+    }
+    props.navigation.goBack();
+    Toast.show('Moved Message');
```
Ah, here's an example of how `Promise#catch` tends to lead to confused error-handling, which is easier to see and avoid when using try/catch.

This code will unconditionally show a toast with "Moved Message", even if the move actually ends up failing.

Instead, we should show a confirmation message like that only after we actually succeed.

gnprice replied:
Both the error message and the success message should be translated. Search for `showErrorAlert` to see examples of how it's done at other call sites. See https://github.com/zulip/zulip-mobile/blob/main/docs/howto/translations.md#all-code-contributors-wiring-our-code-for-translations for background.

gnprice replied:
Write the confirmation message in sentence case: only the first word capitalized, so like "Moved message".

gnprice commented:
Thanks @SilentCruzer for the revision! Comments above. There've also been a few changes upstream which may generate conflicts you'll need to resolve when rebasing: in particular narrows now contain a stream ID instead of a stream name, and the action-sheets code has a more normal structure.

The UI looks pretty reasonable to me from those screenshots and video (thanks!), but I'd be curious if @alya has any feedback to add.

alya commented:
Thanks for picking this up @SilentCruzer !

On the UI, is there a reason to order the parts of the UI differently from the web app? I would put the message content last, unless there's good reason not to.

I guess the notification options are not implemented here? That seems fine for an initial version, but we should make sure to keep track of that as a follow-up when this PR is merged.

alya commented:
The other key part to note is that the permissions are *much* more complicated than described at the top of the PR. This is not an accurate summary:

> Created a new component called MoveMessage to display the details on where to move the message. The screen changes slightly depending on whether the user is admin or not. If the user is an admin, they can change the stream of the message, if not admin they can only change the topic in the current stream.

We may want to wait for https://github.com/zulip/zulip/issues/21739 to be resolved before we finish this feature?

alya commented:
The other visual question I have is whether we have other buttons that look like this? E.g. the buttons in the profile settings UI look like quite different.

gnprice commented:
> The other key part to note is that the permissions are _much_ more complicated than described at the top of the PR. This is not an accurate summary:

Hmmm, good point. I didn't look closely enough and missed that.


> We may want to wait for [zulip/zulip#21739](https://github.com/zulip/zulip/issues/21739) to be resolved before we finish this feature?

Yeah, we probably want to wait for that before attempting to write down the full permissions logic here. Probably also will want the permissions logic to be in shared code -- similar to at https://github.com/zulip/zulip-mobile/pull/5322#issuecomment-1093528762 and the comments after that.

Do you think there would be value in having a version of this feature sooner that, say, was only available when you're an admin? If so, then the permissions logic for just that case may be relatively simple. Then if we merge a version of this UI that we only make available for admins, we can extend it later after we complete that revamp of how these permissions work and also refactor the logic for them into being shared.

alya commented:
> Do you think there would be value in having a version of this feature sooner that, say, was only available when you're an admin? If so, then the permissions logic for just that case may be relatively simple. Then if we merge a version of this UI that we only make available for admins, we can extend it later after we complete that revamp of how these permissions work and also refactor the logic for them into being shared.

Yes, I think merging this feature just for admins and then extending it sounds like a great plan!

SilentCruzer commented:
I have made the suggested changes and updated the PR.

@alya , I have moved the content part to the end and also changed the style of the button to match the other screens.(In the below screenshot, it is the same style used in the 'set a status' screen)

Screenshots:
<img src="https://user-images.githubusercontent.com/73246484/166451208-41b8968b-8bc5-4e0b-b38b-503d4f4ec347.png" width=325  height=700>     <img src="https://user-images.githubusercontent.com/73246484/166451033-9ba71c21-d83d-4a63-85b9-5ef9638d86e0.png" width=325 height=700>

gnprice commented on src/message/MoveMessage.js:
```diff
@@ -0,0 +1,211 @@
+  }),
+};
+
+export default function MoveMessage(props: Props): Node {
+  const themeContext = useContext(ThemeContext);
+  const backgroundColor = themeContext.backgroundColor;
+  const cardColor = themeContext.cardColor;
+  const iconName = Platform.OS === 'android' ? 'arrow-left' : 'chevron-left';
+  const auth = useSelector(getAuth);
+  const allStreams = useSelector(getStreams);
+  const isAdmin = useSelector(getOwnUser).is_admin;
+  const messageId = props.route.params.message.id;
+  const currentStreamId = streamIdOfNarrow(props.route.params.messageNarrow);
+  const currentStreamName = useSelector(state => getStreamForId(state, currentStreamId)).name;
+  const [narrow, setNarrow] = useState(streamNarrow(currentStreamId));
+  const [subject, setSubject] = useState(props.route.params.message.subject);
+  const [propagateMode, setPropagateMode] = useState('change_one');
+  const [streamId, setStreamId] = useState(currentStreamId);
+  const [topicFocus, setTopicFocus] = useState(false);
+  const _ = useContext(TranslationContext);
+
```
nit: Organize this by grouping together different kinds of definitions, with blank lines to separate groups.

Follow the patterns seen in other components in the codebase: first props, then context, then selectors, then state.

gnprice commented on src/message/MoveMessage.js:
```diff
@@ -0,0 +1,211 @@
+  const allStreams = useSelector(getStreams);
+  const isAdmin = useSelector(getOwnUser).is_admin;
+  const messageId = props.route.params.message.id;
+  const currentStreamId = streamIdOfNarrow(props.route.params.messageNarrow);
+  const currentStreamName = useSelector(state => getStreamForId(state, currentStreamId)).name;
+  const [narrow, setNarrow] = useState(streamNarrow(currentStreamId));
+  const [subject, setSubject] = useState(props.route.params.message.subject);
```
These and some other spots would be made a bit simpler to read by pulling the route params out as locals:
```
  const { message, messageNarrow } = props.route.params;
```

That also replaces the need for a `messageId` local -- can just say `message.id`.

gnprice commented on src/message/MoveMessage.js:
```diff
@@ -0,0 +1,211 @@
+
+type Props = $ReadOnly<{|
+  navigation: AppNavigationProp<'move-message'>,
+  route: RouteProp<'move-message', {| message: Message | Outbox, messageNarrow: Narrow |}>,
```
What's the intended meaning of `messageNarrow`? This could use some jsdoc here to clarify.

I'm not sure what it's supposed to mean here. That in turn is necessary in order to properly review the logic below that consumes it.

gnprice reviewed (commented):
Thanks @SilentCruzer for the revision! Comments below.
